libs/core/src/main/java/org/opensearch/core/compress/CompressorRegistry.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.core.compress;
+
+import org.opensearch.common.Nullable;
+import org.opensearch.core.common.bytes.BytesReference;
+import org.opensearch.core.compress.spi.CompressorProvider;
+import org.opensearch.core.xcontent.MediaTypeRegistry;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Objects;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+/**
+ * A registry that wraps a static Map singleton which holds a mapping of unique String names (typically the
+ * compressor header as a string) to registerd {@link Compressor} implementations.
+ *
+ * This enables plugins, modules, extensions to register their own compression implementations through SPI
+ *
+ * @opensearch.internal
+ */
+public final class CompressorRegistry {
+    /** No compression singleton - we still register so users can specify NONE in the API*/
+    public static final Compressor NONE;
+
+    // the backing registry map
+    private static final Map<String, Compressor> registeredCompressors;
+
+    // no instance:
+    private CompressorRegistry() {}
+
+    static {
+        ArrayList<Entry> compressors = new ArrayList<>();
+        for (CompressorProvider provider : ServiceLoader.load(CompressorProvider.class, CompressorProvider.class.getClassLoader())) {
+            compressors.addAll(provider.getCompressors());
+        }
+        registeredCompressors = Map.copyOf(compressors.stream().collect(Collectors.toMap(Entry::getName, Entry::getCompressor)));
+        NONE = registeredCompressors.get(NoneCompressor.NAME);
+    }
+
+    /**
+     * An entry for registering a concrete {@link Compressor} identified by a unique String key
+     *
+     * @opensearch.api
+     * @opensearch.experimental
+     */
+    public static class Entry {
```

Comment:
Do you think we could use `AbstractMap.SimpleEntry<String, Compressor>` here?

Reply:
👍🏻 Yes I went ahead and change it.